lib/puppetserver/ca/list_action.rb

```diff
@@ -0,0 +1,145 @@
+  puppetserver ca list [--config]
+  puppetserver ca list [--all]
+
+Description:
+List outstanding certificate requests. If --all is specified, signed and revoked certificates will be listed as well.
+
+Options:
+      BANNER
+
+      BODY = JSON.dump({desired_state: 'signed'})
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--config CONF', 'Custom path to Puppet\'s config file') do |conf|
+            parsed['config'] = conf
+          end
+          opts.on('--help', 'Display this command specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--all', 'List all certificates') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def run(input)
+        config = input['config']
+
+        if config
+          errors = FileUtilities.validate_file_paths(config)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        puppet = PuppetConfig.parse(config)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        all_certs = get_all_certs(puppet.settings)
+        return 1 if all_certs.nil?
+
+        requested, signed, revoked = separate_certs(all_certs)
+        input['all'] ? human_readable(requested, signed, revoked) : human_readable(requested)
+
+        return 0
+      end
+
+      def human_readable(requested, signed = [], revoked = [])
+
+        if revoked.empty? && signed.empty? && requested.empty?
+          @logger.inform "No certificates to list"
+          return
+        end
+
+        unless requested.empty?
+          @logger.inform "Requested Certificates:"
+          print_output(requested)
+        end
+
+        unless signed.empty?
+          @logger.inform "Signed Certificates:"
+          print_output(signed)
+        end
+
+        unless revoked.empty?
+          @logger.inform "Revoked Certificates:"
+          print_output(revoked)
+        end
+      end
+
+      def print_output(certs)
+        padded = 0
+        certs.each do |cert|
+          cert_size = cert["name"].size
+          padded = cert_size if cert_size > padded
+        end
+
+        certs.each do |cert|
+          @logger.inform "    #{cert["name"]}".ljust(padded + 6) + " (SHA256) #{cert["fingerprints"]["SHA256"]}" +
+                             (cert["dns_alt_names"].empty? ? "" : "\talt names: #{cert["dns_alt_names"]}")
+          end
+      end
+
+      def get_certificate_statuses(settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_statuses', 'any_key')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          connection.get(url)
+        end
+      end
+
+      def separate_certs(all_certs)
+        requested = all_certs.find_all { |v| v["state"] == "requested" }
```

Comment:
I think there's also `Enumerable#sort_by` that will do this in one pass.

Reply:
What do you mean one pass?

Reply:
so if I do ``something = all_certs.sort_by {|v| v["state"]}`` I know that it will sort ``all_certs``  so that each entry is grouped by ``state``, but how do I then split ``something`` into three separate vars? I'm sure there's a way to do this all in one line but I'm not fluent enough in Ruby to figure it out

Reply:
Yeah, i remember doing something similar to that in clojure, but not ruby

Reply:
d'oh. I meant `group_by`, thanks for catching that @amysahli !

Reply:
So when I do ``requested = all_certs.find_all{|v| v["state"] == "requested"}`` and there aren't any, I get ``requested = []``. But if I do ``certs = all_certs.group_by{|v| v["state"]}`` and then ``requested = certs["requested"]`` if there aren't any, I get ``requested = nil``, which breaks everything. Am I doing something wrong?

Reply:
No, that's `group_by`'s API, but we could either change the method signature or  check for nil and return an Array in that case.

So we could create a Hash that returns Arrays for unknown keys rather than nil with this:
```rb
certs = Hash.new([]).merge(all_certs.group_by {|c| c['state'] })
```
I don't love that, it might be a little too clever
Implicit or explicit nil checking might be better:
```rb
certs = all_certs.group_by {|c| c['state'] }
requested = certs.fetch('requested', [])
# or
request = cert['requested'] || []
```

One of those options is probably better than changing the method signature. If we did that we'd then probably have to go back to nil checking in the output method, eg
```rb
def print_output(maybe_requested = nil, maybe_signed = nil, maybe_revoked = nil)
  requested = Array(maybe_requested)
  signed = Array(maybe_signed)
  revoked = Array(maybe_revoked)

  ## ... rest of current method
end
```

Reply:
Okay, I'll start playing around with those and see which works best. Thank you!